src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastCache.cs

```diff
@@ -0,0 +1,461 @@
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ref CastCacheEntry Element(ref int tableData, int index)
+        {
+            // element 0 is used for embedded aux data, skip it
+            return ref Unsafe.Add(ref Unsafe.As<int, CastCacheEntry>(ref tableData), index + 1);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static CastResult TryGet(nuint source, nuint target)
+        {
+            // table is always initialized and is not null.
+            ref int tableData = ref TableData(s_table!);
+
+            int index = KeyToBucket(ref tableData, source, target);
+            for (int i = 0; i < BUCKET_SIZE;)
+            {
+                ref CastCacheEntry pEntry = ref Element(ref tableData, index);
+
+                // we must read in this order: version -> [entry parts] -> version
+                // if version is odd or changes, the entry is inconsistent and thus ignored
+                uint version = Volatile.Read(ref pEntry._version);
+
+#if CORECLR
+                // in CoreCLR we do ordinary reads of the entry parts and
+                // Interlocked.ReadMemoryBarrier() before reading the version
+                nuint entrySource = pEntry._source;
+#else
+                // must read this before reading the version again
+                nuint entrySource = Volatile.Read(ref pEntry._source);
+#endif
+
+                // mask the lower version bit to make it even.
+                // This way we can check if version is odd or changing in just one compare.
+                version &= unchecked((uint)~1);
+
+                if (entrySource == source)
+                {
+
+#if CORECLR
+                    // in CoreCLR we do ordinary reads of the entry parts and
+                    // Interlocked.ReadMemoryBarrier() before reading the version
+                    nuint entryTargetAndResult = pEntry._targetAndResult;
+#else
+                    // must read this before reading the version again
+                    nuint entryTargetAndResult = Volatile.Read(ref pEntry._targetAndResult);
+#endif
+
+                    // target never has its lower bit set.
+                    // a matching entryTargetAndResult would the have same bits, except for the lowest one, which is the result.
+                    entryTargetAndResult ^= target;
+                    if (entryTargetAndResult <= 1)
+                    {
+                        // make sure the second read of 'version' happens after reading 'source' and 'targetAndResults'
+                        //
+                        // We can either:
+                        // - use acquires for both _source and _targetAndResults or
+                        // - issue a load barrier before reading _version
+                        // benchmarks on available hardware (Jan 2020) show that use of a read barrier is cheaper.
+
+#if CORECLR
```

Comment:
If this is required for correctness, it should not be under ifdef.

Reply:
It is either a load fence here or the two loads above need to be acquires. As it was measured, when this was implemented, one fence was noticeably cheaper. 

We need to add `Interlocked.ReadMemoryBarrier()` to NativeAOT. And it needs to be an intrinsic - no point in implementing it as an internal call.

I will log an issue on `Interlocked.ReadMemoryBarrier()`. Once that is added, we can remove ifdefs here and the two loads above.

Reply:
Logged an issue for adding `Interlocked.ReadMemoryBarrier()` intrinsic - https://github.com/dotnet/runtime/issues/84445